Fix XGLM loss computation (PyTorch and TensorFlow)

damianoamatruda opened the pull request:
# What does this PR do?

This PR fixes the loss computation for XGLM in both PyTorch and TensorFlow implementations.

The labels were shifted by one and the padding token was appended, causing artificial loss contributions, inconsistencies between non-padded and right-padded sequences, and potential bias toward predicting padding tokens.

The updated implementations ignore the last logit and do not append the padding token to the labels, aligning with the behavior in GPT-2 and other models.

The logic of the computation was first identified in #22540, where it was ported from the PyTorch implementation to the TensorFlow one for consistency. In this PR I've reverted the TensorFlow implementation to its previous valid behavior and I've updated the PyTorch implementation to match it.

I've also added XGLM tests to ensure that the losses of non-padded and padded inputs match.

This bug was discovered in a joint project while collaborating with @mdrpanwar and @ayushkumartarun.


## Who can review?

@Rocketknight1 @gante @ArthurZucker

damianoamatruda committed:
Fix XGLM loss computation (PyTorch and TensorFlow)

Rocketknight1 commented:
Hi @damianoamatruda, yes, the original code is incorrect! However, a simpler fix would be to change the label padding value to `-100`, which indicates masked positions for our loss computations. Padding/shifting labels is preferable to shifting logits, because logits are a much larger tensor and can carry gradient, so speed + memory usage are affected if we perform lots of operations on them, especially during training.

damianoamatruda commented:
Hi @Rocketknight1, thank you for the clear explanation!

I've updated the PR to shift only the labels, as previously done, and replaced the padding token with the mask value `-100`.

I've also updated the PyTorch test to match the changes introduced in the newly merged PR #35659.

Rocketknight1 reviewed (approved):
Yes, LGTM now! cc @arthurzucker for core maintainer review

Rocketknight1 commented:
run-slow: xglm

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_35878). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Rocketknight1 commented:
run-slow: xglm

Rocketknight1 commented:
Hi @damianoamatruda I'm seeing some failures in the slow tests for XGLM, can you take a look? You can check the CI logs, or to run slow tests locally, you can do something like `RUN_SLOW=1 pytest tests/models/xglm/test_modeling_xglm.py`

damianoamatruda committed:
Update expected output string in XGLM sample test

This updates the expected output string of test_xglm_sample for torch
2.0 to the correct one and removes the one for torch 1.13.1 + cu116
(transformers moved to torch 2.0 with PR #35358).

damianoamatruda committed:
Update expected output IDs in XGLM generation test

Rocketknight1 commented:
Hi @damianoamatruda, I'm not sure exactly what's causing that! It's likely those tests were just flaky on a past commit - can you try rebasing again? If they still won't go away then I'll see if we can actually fix or skip them on `main`.

damianoamatruda commented:
@Rocketknight1, I rebased and the test `tests/models/qwen2_5_vl/test_modeling_qwen2_5_vl.py::Qwen2_5_VLModelTest::test_generate_from_inputs_embeds_1_beam_search` failed again.

Rocketknight1 commented:
Yeah, that test is a problem on `main` right now. Please leave it a couple of days and then check back in - hopefully we'll have resolved it by now!

damianoamatruda commented:
Great!

Cyrilvallez reviewed (commented):
Hey! All LGTM concerning the loss part!

However, I must say that I am skeptical concerning the change in set/get embeddings. It looks like we are changing input type to the functions here (layer vs underlying layer data), which may be breaking for current code. Moreover, the failing test explicitly states that it is expected to fail (and it was never fixed).  
TLDR I'd rather we revert the part on embeddings, and keep the loss part 🤗

damianoamatruda commented:
Hi @Cyrilvallez, done!

The tests now pass without requiring the commits for the embeddings. How did you fix/disable the failing tests?

Thank you for the review.

Cyrilvallez commented:
Thanks for reverting! The failing tests were **slow test** triggered by the github-actions, they are not run by the usual CI which is why you cannot see them now!

damianoamatruda commented:
Is there anything else to do or is everything okay?

ArthurZucker reviewed (approved):
Let's go! @Cyrilvallez has a conference this week hahah sorry 🤗

ArthurZucker commented:
BTW you need to resolve confilcts (probably no changes needed on the modeling non tf side no?

Cyrilvallez commented on src/transformers/models/xglm/modeling_xglm.py:
```diff
@@ -778,10 +778,10 @@ def forward(
         if labels is not None:
-            # shift labels and add a pad token to the end
+            # shift labels to the left and cut last logit token
             shift_labels = labels.new_zeros(labels.shape)
             shift_labels[:, :-1] = labels[:, 1:].clone()
-            shift_labels[:, -1] = self.config.pad_token_id
+            shift_labels[:, -1] = -100
```
Should not be needed anymore, thanks to refactors on main!

damianoamatruda replied:
Refactor #35875 moved the loss computation for XGLM into a dedicated function `xglm_cross_entropy_loss`, but the invalid logic remains: it still appends the padding token to the labels, causing this PR's test `XGLMModelLanguageGenerationTest::test_loss_with_padding` to fail. Removing the dedicated function fixes the issue.

Cyrilvallez commented on tests/models/xglm/test_modeling_xglm.py:
```diff
@@ -424,14 +424,16 @@ def test_xglm_sample(self):
-            # torch 2.0 + cu117
-            "Today is a nice day and the water is still cold. We just stopped off for some fresh",
-        ]
-        self.assertIn(output_str, EXPECTED_OUTPUT_STRS)
+        if torch_device == "cuda":
+            EXPECTED_OUTPUT_STR = (
+                "Today is a nice day and the water is still cold. We just stopped off for some fresh coffee. This place"
+                " looks like a"
+            )
+        else:
+            EXPECTED_OUTPUT_STR = (
+                "Today is a nice day and the sun is shining. A nice day with warm rainy and windy weather today."
+            )
+        self.assertEqual(output_str, EXPECTED_OUTPUT_STR)
 
```
Why did you change this part? Usually generation results change based on the hardware, so this was likely correct on our slow CI hardware

damianoamatruda replied:
I updated the expected output because `XGLMModelLanguageGenerationTest::test_xglm_sample` was failing on the CI hardware, since this produces a slightly longer result. In commit damianoamatruda/transformers@aea148a135e5b1442d24bfc85e7a0dbbf749334a, I changed the expected output from

> "Today is a nice day and the water is still cold. We just stopped off for some fresh"

to

> "Today is a nice day and the water is still cold. We just stopped off for some fresh coffee. This place looks like a"

In commit damianoamatruda/transformers@d602952e448973f1e74b0b47f540477fe9df82bc, I added an alternative expected output for CPU so I could run the test on my own hardware. Since this alternative isn't needed for the CI hardware that you use, I'll remove that commit. Let me know if you agree.

Cyrilvallez replied:
Oh I see, since it does not explicitly set `max_new_tokens`, it must have changed a while back when we switched the default and produce outputs of different length. Thanks!

Cyrilvallez reviewed (commented):
Hey @damianoamatruda! Indeed very sorry, I had a lot going on this week! As you can see, in the meantime `xglm` got the loss refactor incorporated, which automatically fixed the issue at hand in Pytorch. The change to the pytorch modeling should not be needed anymore. Very happy to add the test and the change to tensorflow file though!

damianoamatruda commented:
@ArthurZucker, @Cyrilvallez, no problem, I understand your commitments 🤗

damianoamatruda commented:
Refactor #35875 moved the loss computation for XGLM into a dedicated function `xglm_cross_entropy_loss`, but the invalid logic remains: it still appends the padding token to the labels, causing this PR's test `XGLMModelLanguageGenerationTest::test_loss_with_padding` to fail. Removing the dedicated function fixes the issue.

Cyrilvallez reviewed (approved):
Oh indeed, did not notice that https://github.com/huggingface/transformers/pull/35875 created a dedicated function to ensure BC, but BC was wrong!

All right, LGTM! Thanks a lot for the fix!! 🤗

damianoamatruda commented:
Thank you all, it's been a pleasure for me! 🤗
